opencog/nlp/sureal/SuRealModule.cc

```diff
@@ -0,0 +1,211 @@
+/*
+ * SuRealModule.cc
+ *
+ * Copyright (C) 2014 OpenCog Foundation
+ *
+ * Author: William Ma <https://github.com/williampma>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/atomspace/AtomSpaceUtils.h>
+#include <opencog/guile/SchemePrimitive.h>
+#include <opencog/nlp/types/atom_types.h>
+
+#include "SuRealModule.h"
+#include "SuRealPMCB.h"
+
+
+using namespace opencog::nlp;
+using namespace opencog;
+
+DECLARE_MODULE(SuRealModule);
+
+
+/**
+ * The constructor for SuRealModule.
+ *
+ * @param cs   the OpenCog server
+ */
+SuRealModule::SuRealModule(CogServer& cs) : Module(cs)
+{
+
+}
+
+/**
+ * The required implementation for the pure virtual init method.
+ */
+void SuRealModule::init(void)
+{
+#ifdef HAVE_GUILE
+    define_scheme_primitive("sureal-match", &SuRealModule::do_sureal_match, this);
+    define_scheme_primitive("sureal-get-mapping", &SuRealModule::do_sureal_get_mapping, this);
+#endif
+}
+
+/**
+ * Implement the "sureal-match" scheme primitive.
+ *
+ * Uses the pattern matcher to find all InterpretationNodes whoses
+ * corresponding SetLink contains a structure similar to the input,
+ * taking into accounting each ConceptNode/PredicateNode's corresponding
+ * WordNode and their LG dictionary entry.
+ *
+ * @param h   a SetLink contains the atoms which will become the clauses
+ * @return    a sequence of InterpretationNodes ranked with best result first
+ */
+HandleSeq SuRealModule::do_sureal_match(Handle h)
+{
+#ifdef HAVE_GUILE
+    // clear the old results on new "sureal-match" to sync with "sureal-get-mapping"
+    m_results.clear();
+
+    // only accept SetLink
+    if (h->getType() != SET_LINK)
+        return HandleSeq();
+
+    AtomSpace* pAS = SchemeSmob::ss_get_env_as("sureal-match");
```

Comment:
If you wanted to, you could avoid depending on guile here simply by getting the atomspace from the cogserver. (it would even be a little bit faster).

SchemeSmob::ss_get_env_as() was invented for that code that cannot get access to the cogserver, or equivalently, to avoid having everything to depend on the cogserver.  So you can continue to use it here, if you wish.  You have a choice.